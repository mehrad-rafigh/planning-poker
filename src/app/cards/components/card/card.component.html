<div
  class="card"
  [ngStyle]="{ 'background-color': cardColor.backgroundColor }"
  [ngClass]="setFontColor()"
  (click)="handleClick()"
>
  <span *ngIf="!detailMode" class="top-right quarter-h">{{ value }}</span>
  <span *ngIf="detailMode" class="half-h" [ngClass]="setDetailMode()">{{
    value
  }}</span>
  <span *ngIf="!detailMode" class="bottom-left quarter-h">{{ value }}</span>
</div>
