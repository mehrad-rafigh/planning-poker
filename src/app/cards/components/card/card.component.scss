@import "../../../../mixins";

span {
  font-weight: bold;
}

.card {
  display: flex;
  flex-flow: column nowrap;
  box-sizing: border-box;
  height: 200px;
  width: auto;
  border-color: lightslategray;
  position: relative;
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);
  transition: 0.3s;
  border-radius: 15px;

  @include mobile() {
    height: 250px;
    font-size: 26px;
  }

  @include tablet() {
    height: 300px;
    font-size: 24px;
  }

  @include above-tablet() {
    height: 350px;
    font-size: 22px;
  }
}

.top-right {
  text-align: right;
  margin-top: 32px;
  margin-right: 32px;
}

.bottom-left {
  margin-bottom: 32px;
  margin-left: 32px;
  position: absolute;
  bottom: 0;
  height: auto !important;
}

.quarter-h {
  height: 25%;
}

.half-h {
  height: 50%;
}

.detail-mode {
  height: unset;
  margin: auto;
  font-size: 128px;
}

.font-white {
  color: white;
}
