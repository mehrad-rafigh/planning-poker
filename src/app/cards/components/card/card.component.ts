import {
  Component,
  OnInit,
  ChangeDetectionStrategy,
  Input,
  EventEmitter,
  Output
} from "@angular/core";

@Component({
  selector: "app-card",
  templateUrl: "./card.component.html",
  styleUrls: ["./card.component.scss"],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class CardComponent implements OnInit {
  @Input() value: string;
  @Input() cardColor: { backgroundColor: string; fontColor: string };
  @Input() detailMode = false;
  @Output() clickEvent = new EventEmitter<any>();
  constructor() {}

  ngOnInit(): void {}

  handleClick(): void {
    this.clickEvent.emit({
      value: this.value,
      card: {
        backgroundColor: this.cardColor.backgroundColor,
        fontColor: this.cardColor.fontColor
      }
    });
  }

  setDetailMode(): string {
    if (this.detailMode) {
      return "detail-mode";
    }
  }

  setFontColor(): string {
    if (this.cardColor.fontColor === "white") {
      return "font-white";
    }
  }
}
