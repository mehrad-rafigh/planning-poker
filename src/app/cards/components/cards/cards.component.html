<app-toolbar></app-toolbar>
<div class="cards-container">
  <app-card
    *ngFor="let no of range"
    [value]="no"
    [cardColor]="setCardColors()"
    (clickEvent)="showCardInDetail($event)"
  ></app-card>
</div>
