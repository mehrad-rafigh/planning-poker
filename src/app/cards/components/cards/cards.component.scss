@import "../../../../mixins";

.cards-container {
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-template-rows: repeat(3, 1fr);
  grid-gap: 16px;
  margin: 16px;

  @include mobile() {
    grid-template-columns: repeat(2, 1fr);
  }

  @include tablet() {
    grid-template-columns: repeat(3, 1fr);
  }

  @include above-tablet() {
    grid-template-columns: repeat(4, 1fr);
  }
}
