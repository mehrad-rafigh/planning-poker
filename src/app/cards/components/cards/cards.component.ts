import {
  Component,
  OnInit,
  ChangeDetectionStrategy,
  Input
} from "@angular/core";
import { Router } from "@angular/router";

@Component({
  selector: "app-cards",
  templateUrl: "./cards.component.html",
  styleUrls: ["./cards.component.scss"],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class CardsComponent implements OnInit {
  @Input() range: string[] = [
    "0",
    "0.5",
    "1",
    "2",
    "3",
    "5",
    "8",
    "13",
    "21",
    "34",
    "55",
    "89",
    "144",
    "?",
    "\u2615",
    "üÉè"
  ];
  constructor(private router: Router) {}

  ngOnInit(): void {}

  showCardInDetail($event): void {
    this.router.navigate(["detail"], {
      state: {
        value: $event.value,
        card: {
          backgroundColor: $event.card.backgroundColor,
          fontColor: "white"
        }
      }
    });
  }

  setCardColors(): { backgroundColor: string; fontColor: string } {
    const backgroundColor = this.calculateRandomRGBColor();
    const fontColor = this.shouldFontBeWhite(backgroundColor)
      ? "white"
      : "black";

    return { backgroundColor, fontColor };
  }

  private calculateRandomRGBColor(): string {
    const red = Math.ceil(Math.random() * 255);
    const green = Math.ceil(Math.random() * 255);
    const blue = Math.ceil(Math.random() * 255);
    return `rgb(${red},${green},${blue})`;
  }

  private shouldFontBeWhite(backgroundColor: string): boolean {
    const sp = backgroundColor
      .replace("rgb", "")
      .replace("(", "")
      .replace(")", "")
      .split(",");

    const red = Number(sp[0]);
    const blue = Number(sp[1]);
    const green = Number(sp[2]);

    return red < 100 || blue < 100 || green < 100;
  }
}
