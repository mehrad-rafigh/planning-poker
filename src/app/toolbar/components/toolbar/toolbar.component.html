<nav class="navbar d-flex row-no-wrap">
  <span class="navbar-title f-start">Planning Poker</span>
</nav>
