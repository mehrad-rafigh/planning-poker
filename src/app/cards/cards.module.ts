import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { CardsComponent } from "./components/cards/cards.component";
import { ToolbarModule } from "../toolbar/toolbar.module";
import { CardComponent } from "./components/card/card.component";
import { DetailComponent } from "./components/detail/detail.component";

@NgModule({
  declarations: [CardsComponent, CardComponent, DetailComponent],
  imports: [CommonModule, ToolbarModule],
  exports: [CardsComponent, DetailComponent]
})
export class CardsModule {}
