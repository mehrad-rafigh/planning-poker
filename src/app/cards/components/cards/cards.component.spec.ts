import { async, ComponentFixture, TestBed } from "@angular/core/testing";

import { CardsComponent } from "./cards.component";
import { RouterTestingModule } from "@angular/router/testing";
import { DebugElement } from "@angular/core";
import { Router } from "@angular/router";
import { CardComponent } from "../card/card.component";
import { By } from "@angular/platform-browser";
import { ToolbarComponent } from "../../../toolbar/components/toolbar/toolbar.component";
import { compilePipeFromMetadata } from "@angular/compiler";

describe("CardsComponent", () => {
  let component: CardsComponent;
  let fixture: ComponentFixture<CardsComponent>;
  let debugElement: DebugElement;
  let router: Router;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [RouterTestingModule],
      declarations: [CardsComponent, CardComponent, ToolbarComponent]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CardsComponent);
    component = fixture.componentInstance;
    component.range = [
      "0",
      "0.5",
      "1",
      "2",
      "3",
      "5",
      "8",
      "13",
      "21",
      "34",
      "55",
      "89",
      "144",
      "?",
      "\u2615",
      "🃏"
    ];
    debugElement = fixture.debugElement;
    router = TestBed.inject(Router);

    component.calculateRandomRGBColor = jest
      .fn()
      .mockImplementation(() => "rgb(116, 254, 117)");

    fixture.detectChanges();
  });

  it("should create", () => {
    expect(component).toBeTruthy();
  });

  it("should match the snapshot", () => {
    expect(fixture).toMatchSnapshot();
  });
});
