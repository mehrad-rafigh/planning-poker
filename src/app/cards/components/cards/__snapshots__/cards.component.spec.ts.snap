// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardsComponent should match the snapshot 1`] = `
<app-cards
  calculateRandomRGBColor={[Function Function]}
  range={[Function Array]}
  router={[Function Router]}
>
  <app-toolbar>
    <nav
      class="navbar d-flex row-no-wrap"
    >
      <span
        class="navbar-title f-start"
      >
        Planning Poker
      </span>
    </nav>
  </app-toolbar><div
    class="cards-container"
  >
    
    <app-card>
      <div
        class="card font-white"
        style="background-color: rgb(116, 254, 117);"
      >
        
        <span
          class="top-right quarter-h"
        >
          0
        </span>
        
        
        <span
          class="bottom-left quarter-h"
        >
          0
        </span>
      </div>
    </app-card>
    <app-card>
      <div
        class="card font-white"
        style="background-color: rgb(116, 254, 117);"
      >
        
        <span
          class="top-right quarter-h"
        >
          0.5
        </span>
        
        
        <span
          class="bottom-left quarter-h"
        >
          0.5
        </span>
      </div>
    </app-card>
    <app-card>
      <div
        class="card font-white"
        style="background-color: rgb(116, 254, 117);"
      >
        
        <span
          class="top-right quarter-h"
        >
          1
        </span>
        
        
        <span
          class="bottom-left quarter-h"
        >
          1
        </span>
      </div>
    </app-card>
    <app-card>
      <div
        class="card font-white"
        style="background-color: rgb(116, 254, 117);"
      >
        
        <span
          class="top-right quarter-h"
        >
          2
        </span>
        
        
        <span
          class="bottom-left quarter-h"
        >
          2
        </span>
      </div>
    </app-card>
    <app-card>
      <div
        class="card font-white"
        style="background-color: rgb(116, 254, 117);"
      >
        
        <span
          class="top-right quarter-h"
        >
          3
        </span>
        
        
        <span
          class="bottom-left quarter-h"
        >
          3
        </span>
      </div>
    </app-card>
    <app-card>
      <div
        class="card font-white"
        style="background-color: rgb(116, 254, 117);"
      >
        
        <span
          class="top-right quarter-h"
        >
          5
        </span>
        
        
        <span
          class="bottom-left quarter-h"
        >
          5
        </span>
      </div>
    </app-card>
    <app-card>
      <div
        class="card font-white"
        style="background-color: rgb(116, 254, 117);"
      >
        
        <span
          class="top-right quarter-h"
        >
          8
        </span>
        
        
        <span
          class="bottom-left quarter-h"
        >
          8
        </span>
      </div>
    </app-card>
    <app-card>
      <div
        class="card font-white"
        style="background-color: rgb(116, 254, 117);"
      >
        
        <span
          class="top-right quarter-h"
        >
          13
        </span>
        
        
        <span
          class="bottom-left quarter-h"
        >
          13
        </span>
      </div>
    </app-card>
    <app-card>
      <div
        class="card font-white"
        style="background-color: rgb(116, 254, 117);"
      >
        
        <span
          class="top-right quarter-h"
        >
          21
        </span>
        
        
        <span
          class="bottom-left quarter-h"
        >
          21
        </span>
      </div>
    </app-card>
    <app-card>
      <div
        class="card font-white"
        style="background-color: rgb(116, 254, 117);"
      >
        
        <span
          class="top-right quarter-h"
        >
          34
        </span>
        
        
        <span
          class="bottom-left quarter-h"
        >
          34
        </span>
      </div>
    </app-card>
    <app-card>
      <div
        class="card font-white"
        style="background-color: rgb(116, 254, 117);"
      >
        
        <span
          class="top-right quarter-h"
        >
          55
        </span>
        
        
        <span
          class="bottom-left quarter-h"
        >
          55
        </span>
      </div>
    </app-card>
    <app-card>
      <div
        class="card font-white"
        style="background-color: rgb(116, 254, 117);"
      >
        
        <span
          class="top-right quarter-h"
        >
          89
        </span>
        
        
        <span
          class="bottom-left quarter-h"
        >
          89
        </span>
      </div>
    </app-card>
    <app-card>
      <div
        class="card font-white"
        style="background-color: rgb(116, 254, 117);"
      >
        
        <span
          class="top-right quarter-h"
        >
          144
        </span>
        
        
        <span
          class="bottom-left quarter-h"
        >
          144
        </span>
      </div>
    </app-card>
    <app-card>
      <div
        class="card font-white"
        style="background-color: rgb(116, 254, 117);"
      >
        
        <span
          class="top-right quarter-h"
        >
          ?
        </span>
        
        
        <span
          class="bottom-left quarter-h"
        >
          ?
        </span>
      </div>
    </app-card>
    <app-card>
      <div
        class="card font-white"
        style="background-color: rgb(116, 254, 117);"
      >
        
        <span
          class="top-right quarter-h"
        >
          ‚òï
        </span>
        
        
        <span
          class="bottom-left quarter-h"
        >
          ‚òï
        </span>
      </div>
    </app-card>
    <app-card>
      <div
        class="card font-white"
        style="background-color: rgb(116, 254, 117);"
      >
        
        <span
          class="top-right quarter-h"
        >
          üÉè
        </span>
        
        
        <span
          class="bottom-left quarter-h"
        >
          üÉè
        </span>
      </div>
    </app-card>
  </div>
</app-cards>
`;
